Fee Estimator updates from Validation Interface/CScheduler thread

ismaelsadeeq opened the pull request:
This is an attempt to  #11775 

This Pr will enable fee estimator to listen to ValidationInterface notifications to process new transactions added and removed from the mempool.

This PR includes the following changes:

- Added a new callback to the Validation Interface `MempoolTransactionsRemovedForConnectedBlock`, which notifies listeners about the transactions that have been removed due to a new block being connected, along with the height at which the transactions were removed.
- Modified the `TransactionAddedToMempool` callback parameter to include additional information about the transaction needed for fee estimation.
- Updated `CBlockPolicyEstimator` to process transactions using` CTransactionRef` instead of `CTxMempoolEntry.`
- Implemented the `CValidationInterface` interface in `CBlockPolicyEstimater` and overridden the `TransactionAddedToMempool`, `TransactionRemovedFromMempool`, and `MempoolTransactionsRemovedForConnectedBlock` methods to receive updates from their notifications.

Prior to this PR, the fee estimator updates from the mempool, i.e whenever a new block is connected all transactions in the block that are in our mempool are going to be removed using the `removeForBlock` function in `txmempool.cpp`. 

This removal triggered updates to the fee estimator. As a result, the fee estimator would block mempool's `cs` until it finished updating every time a new block was connected.
Instead of being blocked only on mempool tx removal, we were blocking on both tx removal and fee estimator updating.
If we want to further improve fee estimation, or add heavy-calulation steps to it, it is currently not viable as we would be slowing down block relay in the process

This PR is smaller in terms of the changes made compared to #11775, as it focuses solely on enabling fee estimator updates from the validationInterface/cscheduler thread notifications.

I have not split the validation interface because, as I understand it, the rationale behind the split in #11775 was to have `MempoolInterface` signals come from the mempool and `CValidationInterface` events come from validation. I believe this separation can be achieved in a separate refactoring PR when the need arises.

Also left out some commits from #11775
- Some refactoring which are no longer needed.
- Handle reorgs much better in fee estimator. 
- Track witness hash malleation in fee estimator 

I believe they are a separate change that can come in a follow-up after this.

willcl-ark commented:
Nice work!

So if correctly implemented this should speed up block relay, and enable (future) fee estimator changes to be added without blocking critical codepaths.

It would be nice to try and benchmark this somehow, but I'm unsure immediately what the best way to do this would be...

@glozow @TheCharlatan @instagibbs would be curious to know what y'all think of the approach here.

vincenzopalazzo commented on src/init.cpp:
```diff
@@ -348,6 +348,10 @@ void Shutdown(NodeContext& node)
     node.scheduler.reset();
 
+    if (node.fee_estimator) {
+        UnregisterValidationInterface(node.fee_estimator.get());
```
```suggestion
        UnregisterSharedValidationInterface(node.fee_estimator);
```

if you want you can also do the check of null pointer inside the `UnregisterValidationInterface`

ismaelsadeeq replied:
Thanks, I will leave this as is since there is a check above.

vincenzopalazzo commented on src/policy/fees.cpp:
```diff
@@ -605,36 +591,39 @@ void CBlockPolicyEstimator::processTransaction(const CTxMemPoolEntry& entry, boo
-    feeStats->Record(blocksToConfirm, (double)feeRate.GetFeePerK());
-    shortStats->Record(blocksToConfirm, (double)feeRate.GetFeePerK());
-    longStats->Record(blocksToConfirm, (double)feeRate.GetFeePerK());
+    feeStats->Record(blocksToConfirm, (double)feerate);
+    shortStats->Record(blocksToConfirm, (double)feerate);
+    longStats->Record(blocksToConfirm, (double)feerate);
```
I will also change this in a separate commit to use static cast with CPP std

ismaelsadeeq replied:
Fixed thanks.

vincenzopalazzo reviewed (commented):
I did a first pass on this but I should take a look in a more deep way

glozow commented:
Concept ACK! Background fee estimator would be really nice (fwiw it seems like this was always the plan: https://github.com/bitcoin/bitcoin/pull/10199#discussion_r115530273 and  #11775 had a lot of fans). Updating asynchronously instead of blocking `removeForBlock` / `ConnectTip` should make things faster - I personally don't need a bench to be convinced though it'd be nice to see. I don't think `CTxMemPool` should know anything about there being a fee estimator; the dependency should be the other way around. I agree that if we want to pursue having multiple fee estimator approaches (e.g. #27995) it would be best for them to be in the background so we don't need to worry about the performance impact on every mempool operation. I can also imagine using this to test fee estimator PRs, as it'd be easy to create 2 fee estimators that use the exact same mempool data and compare their results.

I haven't reviewed the new interface closely yet but will do as soon as I can.

darosior commented:
Concept ACK. Not sure i'll be able to review anytime soon though.

glozow commented on src/txmempool.cpp:
```diff
@@ -481,13 +481,16 @@ void CTxMemPool::removeUnchecked(txiter it, MemPoolRemovalReason reason)
         GetMainSignals().TransactionRemovedFromMempool(it->GetSharedTx(), reason, mempool_sequence);
+        if (minerPolicyEstimator) {
+            minerPolicyEstimator->removeTx(hash, false);
+        }
```
4f22c0da092e1c2287e4bc2d84e646d56abbe133
Why move this into the `reason != BLOCK` case?

ismaelsadeeq replied:
Any transaction which reason is `BLOCK` will be captured by `MempoolBlockConnect`

glozow replied:
Ok, I think this should be its own commit with this explanation as it is not relevant to refactoring `processBlock`

glozow reviewed (commented):
Did a first pass

ismaelsadeeq committed:
tx fees, policy: cast with static_cast instead of C-Style cast

ismaelsadeeq commented on src/txmempool.cpp:
```diff
@@ -618,33 +611,25 @@ void CTxMemPool::removeConflicts(const CTransaction &tx)
 
 /**
- * Called when a block is connected. Removes from mempool and updates the miner fee estimator.
+ * Called when a block is connected. Removes from mempool`  .
```
Updated thank you.

ismaelsadeeq commented on src/txmempool.cpp:
```diff
@@ -387,7 +387,6 @@ void CTxMemPoolEntry::UpdateAncestorState(int32_t modifySize, CAmount modifyFee,
 
 CTxMemPool::CTxMemPool(const Options& opts)
     : m_check_ratio{opts.check_ratio},
-      minerPolicyEstimator{opts.estimator},
```
I remove the include thank you.

ismaelsadeeq commented on src/txmempool.cpp:
```diff
@@ -634,17 +634,20 @@ void CTxMemPool::removeForBlock(const std::vector<CTransactionRef>& vtx, unsigne
     }
     // Before the txs in the new block have been removed from the mempool, update policy estimates
     if (minerPolicyEstimator) {minerPolicyEstimator->processBlock(nBlockHeight, entries);}
+    std::vector<CTransactionRef> txs_removed_for_block;
```
Fixed thank you

ismaelsadeeq commented on src/test/fuzz/tx_pool.cpp:
```diff
@@ -59,9 +60,9 @@ struct TransactionsDelta final : public CValidationInterface {
         : m_removed{r}, m_added{a} {}
 
-    void TransactionAddedToMempool(const CTransactionRef& tx, uint64_t /* mempool_sequence */) override
+    void TransactionAddedToMempool(const NewMempoolTransactionInfo& tx_info, uint64_t /* mempool_sequence */) override
```
I remove it out of the struct, thank you.

ismaelsadeeq commented on src/validation.cpp:
```diff
@@ -1196,7 +1196,14 @@ bool MemPoolAccept::SubmitPackage(const ATMPArgs& args, std::vector<Workspace>&
                                              ws.m_base_fees, effective_feerate, effective_feerate_wtxids));
-            GetMainSignals().TransactionAddedToMempool(ws.m_ptx, m_pool.GetAndIncrementSequence());
+            const CTransaction& tx = *ws.m_ptx;
+            bool validForFeeEstimation = !args.m_bypass_limits && !args.m_package_submission && IsCurrentForFeeEstimation(m_active_chainstate) && m_pool.HasNoInputsOf(tx);
```
I updated and now the fee estimator decides, but I did that in the last commit, I dont want to touch `addUnchecked` parameters in this PR.
But I will do a follow up to clean it up and remove `validForFeeEstimate` after this PR.

ismaelsadeeq commented on src/policy/fees.cpp:
```diff
@@ -566,37 +566,28 @@ CBlockPolicyEstimator::CBlockPolicyEstimator(const fs::path& estimation_filepath
-        // It will be synced next time a block is processed.
-        return;
-    }
-
```
This was from rebase from a734d1d166f9936e453b3c1f65b7a569af894082
I think it has to do with [Handle reorgs much better in fee estimator.](https://github.com/bitcoin/bitcoin/pull/11775/commits/660190c8f52737bb26b99d05eace4b5d754565c1)

I think its worth to be in its own PR, its not part of this.

I am removing it from this PR.

TheCharlatan commented on src/policy/fees.cpp:
```diff
@@ -603,36 +603,39 @@ void CBlockPolicyEstimator::processTransaction(const CTransactionRef& tx, const
+
+    // Feerates are stored and reported as BTC-per-kb:
+    CAmount feerate = pos->second.m_fee_per_k;
+    _removeTx(tx->GetHash(), true);
```
Why is this line moved?

ismaelsadeeq replied:
Because `_removeTx` clear the tx from `mapMemPoolTxs`.
So using
```c++
 _removeTx(tx->GetHash(), true);
 std::map<uint256, TxStatsInfo>::iterator pos = mapMemPoolTxs.find(tx->GetHash());
```  
`pos` will always be an empty iterator.

So I find the iterator of the tx from `mapMemPoolTxs` and capture all the values I need (tx fee rate and entry height) before removing the tx. 

Before this pull request, we could get all those details from CTxMempoolEntry.

TheCharlatan replied:
Thanks for the explainer :)
Feel free to resolve.

ismaelsadeeq commented on src/validationinterface.h:
```diff
@@ -130,9 +131,18 @@ class CValidationInterface {
+     *
+     * Called on a background thread.
+     */
+    virtual void MempoolBlockConnect(const std::vector<CTransactionRef>& txs_removed_for_block, unsigned int nBlockHeight) {}
```
Thank you for your review.
I updated the name

ismaelsadeeq commented on src/init.cpp:
```diff
@@ -348,6 +348,10 @@ void Shutdown(NodeContext& node)
     node.chainman.reset();
     node.scheduler.reset();
 
+    if (node.fee_estimator) {
```
Good catch.
I moved it below the Flush

ismaelsadeeq commented on src/validationinterface.h:
```diff
@@ -7,6 +7,7 @@
 #define BITCOIN_VALIDATIONINTERFACE_H
 
 #include <kernel/cs_main.h>
+#include <kernel/mempool_entry.h>
```
Thanks fixed

TheCharlatan commented:
Approach ACK

Will give this another pass soon.

ismaelsadeeq commented:
Force pushed to 55beb4f39bb029e19b333dea205e4b035d01bd12

- Modified `MempoolTransactionsRemovedForConnectedBlock` callback parameter to vector of `NewMempoolTransactionInfo` instead of just CTransaction to resolve conflict with #25380.
See [comment](https://github.com/bitcoin/bitcoin/pull/25380#issuecomment-1713512238)
- Hence add `CTransactionRef` back to `NewMempoolTransactionInfo`

vincenzopalazzo commented on src/kernel/mempool_entry.h:
```diff
@@ -170,10 +170,43 @@ class CTxMemPoolEntry
+struct NewMempoolTransactionInfo {
+    CTransactionRef m_tx;
+    std::vector<CTransactionRef> m_parents;
+    //! The fee the added transaction paid
```
why not
```suggestion
    /* The fee the added transaction paid */
```

vincenzopalazzo reviewed (commented):
Sorry if this get me so long

Concept ACK https://github.com/bitcoin/bitcoin/pull/28368/commits/2d4b66b3b9fd907387fe96676c025b54d9925ebb

I should test it but to me looks good, I had a similar code in my branch :)

TheCharlatan commented on src/kernel/mempool_entry.h:
```diff
@@ -170,6 +170,15 @@ class CTxMemPoolEntry
     const Parents& GetMemPoolParentsConst() const { return m_parents; }
```
In commit 79bcc5ca0679daf6e57fc4d7ce2244262a7cfd13 the last paragraph of its commit message seems like it misses part of a sentence.

ismaelsadeeq replied:
Updated the commit message

sedited reviewed (approved):
ACK 79bcc5ca0679daf6e57fc4d7ce2244262a7cfd13

ismaelsadeeq commented on src/test/fuzz/package_eval.cpp:
```diff
@@ -20,6 +20,8 @@
 
 using node::NodeContext;
 
+struct NewMempoolTransactionInfo;
```
Thank you for you review, removed the forward declaration.

ismaelsadeeq commented on src/policy/fees.cpp:
```diff
@@ -566,37 +566,36 @@ CBlockPolicyEstimator::CBlockPolicyEstimator(const fs::path& estimation_filepath
     LOCK(m_cs_fee_estimator);
-    unsigned int txHeight = entry.GetHeight();
-    uint256 hash = entry.GetTx().GetHash();
+    const CTransactionRef& tx = tx_info.m_tx;
```
Removed, since I can get the hash directly with `tx_info.m_tx->GetHash();`

ismaelsadeeq commented on src/test/policyestimator_tests.cpp:
```diff
@@ -67,16 +67,16 @@ BOOST_AUTO_TEST_CASE(BlockPolicyEstimates)
-                         true},
-                        mpool.GetAndIncrementSequence());
+                    NewMempoolTransactionInfo tx_info;
+                    tx_info.m_tx = MakeTransactionRef(tx);
```
Fixed, thank you

pablomartin4btc reviewed (commented):
Concept ACK

I agree with @glozow regarding the correct objs dependency direction (fee estimator-> `CTxMemPool`) and that this would be very useful in order to test fee estimator PRs.

I'm trying to finish a code review and perform some testing before the [Review club](https://bitcoincore.reviews/28368) in the next 2 days. In the meantime, could the label be added to this PR?

TheCharlatan commented on src/test/fuzz/policy_estimator.cpp:
```diff
@@ -56,12 +70,21 @@ FUZZ_TARGET(policy_estimator, .init = initialize_policy_estimator)
+                        .m_virtual_transaction_size = mempool_entry.GetTxSize(),
+                        .txHeight = mempool_entry.GetHeight(),
+                        .nSizeWithAncestors = mempool_entry.GetSizeWithAncestors(),
+                        .nModFeesWithAncestors = mempool_entry.GetModFeesWithAncestors(),
```
Nitty nit: Inconsistent with closing bracket placement.

TheCharlatan commented on src/test/fuzz/policy_estimator.cpp:
```diff
@@ -41,7 +42,20 @@ FUZZ_TARGET(policy_estimator, .init = initialize_policy_estimator)
                 const CTransaction tx{*mtx};
-                block_policy_estimator.processTransaction(ConsumeTxMemPoolEntry(fuzzed_data_provider, tx), fuzzed_data_provider.ConsumeBool());
+                const CTxMemPoolEntry entry = ConsumeTxMemPoolEntry(fuzzed_data_provider, tx);
+                NewMempoolTransactionInfo tx_info = {
```
Nit: There is trailing whitespace here in commit 5ab08d23b2e88c1a8fda1535114e04b695c3483b. Also, why remove the const qualifier? Can't all the `tx_info` be made `const`?

ismaelsadeeq replied:
Thank you, Will update if there is need to re touch.

sedited reviewed (approved):
Re-ACK 59db4d1739de1a008e6f34d0f03c3e8e8da60a93

hernanmarino reviewed (commented):
Approach ACK . I'm really curious about benchmarks but I don't think they are really necessary to make a decision about this. I also believe this PR will have a higher impact in the future, should any changes to fee estimation be necessary. I 'll try to take a deeper look and code review in a couple of days, if still unmerged.

TheCharlatan commented:
Seems like with the changes here, the `policy/fees` can be removed the kernel library. If you want to, you can integrate this [one-line commit](https://github.com/TheCharlatan/bitcoin/commit/32846141df1a2e110f84f3244d5e98626d379185), otherwise I'll make a tiny follow up.

glozow commented on src/kernel/mempool_entry.h:
```diff
@@ -191,6 +191,10 @@ struct NewMempoolTransactionInfo {
      */
     int64_t m_virtual_transaction_size;
     unsigned int txHeight;
+    bool m_from_disconnected_block;
+    bool m_submitted_in_package;
+    bool m_chainstate_is_current;
+    bool m_has_no_mempool_parents;
```
9b0843042ab5c30e6ef4753d590c69c4811d9a70 seems to be doing multiple things: (1) making `CBlockPolicyEstimator` react to validationinterface and (2) delegating the `validForFeeEstimation` from validation to the `CBlockPolicyEstimator` by providing these bools in the `TransactionAddedToMempool` signal (but this isn't fully done until the next commit which cleans things up).

These should be separate commits, e.g. (2) and then (1).

ismaelsadeeq replied:
Thank you

This is fixed.
1.  a469d937dbbadedeb58e0d5d1c44ed7881a7ebb9 introduces `NewMempoolTransactionInfo` with all the necessary fields and updates `TransactionAddedToMempool` callback parameter.
2. ac82f58e45421aa5aec1d3e6e992d1c369472a10 updates the fee estimator from `CValidationInterface` notifications and manage delegating `validForFeeEstimation`  to `CBlockPolicyEstimator::processTransaction` together the cleanups.

glozow commented on src/validation.cpp:
```diff
@@ -1214,11 +1214,16 @@ bool MemPoolAccept::SubmitPackage(const ATMPArgs& args, std::vector<Workspace>&
+            .m_from_disconnected_block = args.m_bypass_limits,
+            .m_submitted_in_package = args.m_package_submission,
+            .m_chainstate_is_current = IsCurrentForFeeEstimation(m_active_chainstate),
+            .m_has_no_mempool_parents = m_pool.HasNoInputsOf(tx)};
```
This is quite verbose and repeated in several places. Why not create a dedicated constructor?

ismaelsadeeq replied:
Done

glozow commented on src/validationinterface.h:
```diff
@@ -131,9 +131,18 @@ class CValidationInterface {
+     *
+     * Called on a background thread.
+     */
+    virtual void MempoolTransactionsRemovedForConnectedBlock(const std::vector<CTransactionRef>& txs_removed_for_block, unsigned int nBlockHeight) {}
```
This function is newly added in 4986edb99f8aa73f72e87f3bdc09387c3e516197 and then changed in 863d46751900a0c69c4f7fbd92d4d8a651b2cc7b

ismaelsadeeq replied:
Fixed

glozow commented on src/test/fuzz/policy_estimator.cpp:
```diff
@@ -68,10 +70,18 @@ FUZZ_TARGET(policy_estimator, .init = initialize_policy_estimator)
                     mempool_entries.push_back(ConsumeTxMemPoolEntry(fuzzed_data_provider, tx));
                 }
-                std::vector<CTransactionRef> txs;
+                std::vector<const NewMempoolTransactionInfo> txs;
```
863d46751900a0c69c4f7fbd92d4d8a651b2cc7b why are you making the element type const?

ismaelsadeeq replied:
This is fixed in latest push.

glozow commented on src/kernel/mempool_entry.h:
```diff
@@ -191,6 +201,8 @@ struct NewMempoolTransactionInfo {
      */
     int64_t m_virtual_transaction_size;
     unsigned int txHeight;
+    int64_t nSizeWithAncestors;
+    CAmount nModFeesWithAncestors;
```
These fields added in 53bbc791d47a51a7c2bbefb32835fd4001b68890 aren't necessary as you're not using them anywhere. At quick glance I don't think we should ever tell the fee estimator about *modified* fees.

glozow reviewed (commented):
Approach ACK, but some of the changes seem to be interleaved across multiple commits / lumped together in a single commit. For example, `processBlock`'s `txs_removed_for_block` param changes type 3 times in this PR. In an intermediate step where you don't have fee information for a tx, you add a `TxStatsInfo::CAmount` to store it, but then don't delete it at the end when we don't need it anymore.

I'd suggest making this PR 1 change per commit, and 1 commit per change:
- move `removeTx` into the `reason != BLOCK` condition
- remove C-style casts
- Introduce the `RemovedMempoolTransactionInfo` struct and `MempoolTransactionsRemovedForBlock` callback
- Update the fee estimator interface to use `RemovedMempoolEntry`
- Introduce a `NewMempoolTransactionInfo` struct and re-delegate calculation of `validForFeeEstimation` from validation to fee estimator (including the cleanups)
- Make the fee estimator a client of `CValidationInterface`

ismaelsadeeq committed:
tx fees, policy: bugfix: move `removeTx` into reason != `BLOCK` condition

If the removal reason of a transaction is BLOCK, then the `removeTx`
boolean argument should be true.

Before this PR, `CBlockPolicyEstimator` have to complete updating the fee stats
before the mempool clears that's why having removeTx call outside reason!= `BLOCK`
in `addUnchecked` was not a bug.

But in a case where the `CBlockPolicyEstimator` update is asynchronous, the mempool might
clear before we update the `CBlockPolicyEstimator` fee stats.
Transactions that are removed for `BLOCK` reasons will also be incorrectly removed from
`CBlockPolicyEstimator` stats as failures.

ismaelsadeeq committed:
CValidationInterface, mempool: add new callback to `CValidationInterface`

This commit adds a new callback `MempoolTransactionsRemovedForBlock` which notify
its listeners of the transactions that are removed from the mempool because a new
block is connected, along with the block height the transactions were removed.
The transactions are in `RemovedMempoolTransactionInfo` format.

`CTransactionRef`, base fee, virtual size, and height which the transaction was added
to the mempool are all members of the struct called `RemovedMempoolTransactionInfo`.

A struct `NewMempoolTransactionInfo`, which has fields similar to `RemovedMempoolTransactionInfo`,
will be added in a later commit, create a struct `TransactionInfo` with all similar fields.
They can both have a member with type `TransactionInfo`.

ismaelsadeeq committed:
tx fees, policy: update `CBlockPolicyEstimator::processBlock` parameter

Update `processBlock` parameter to reference to a vector of `RemovedMempoolTransactionInfo`.

ismaelsadeeq committed:
CValidationInterface: modify the parameter of `TransactionAddedToMempool`

Create a new struct `NewMempoolTransactionInfo` that will be used as the new parameter of
`TransactionAddedToMempool` callback.

ismaelsadeeq committed:
tx fees, policy: CBlockPolicyEstimator update from `CValidationInterface` notifications

`CBlockPolicyEstimator` will implement `CValidationInterface` and
subscribe to its notification to process transactions added and removed
from the mempool.

Re-delegate calculation of `validForFeeEstimation` from validation to fee estimator.

Also clean up the validForFeeEstimation arg thats no longer needed in `CTxMempool`.

Co-authored-by: Matt Corallo <git@bluematt.me>

ismaelsadeeq commented on src/validationinterface.h:
```diff
@@ -131,9 +132,18 @@ class CValidationInterface {
-     * Provides a vector of transactions evicted from the mempool as a result.
+     * Provides the block that was connected.
+     * The block contains a vector of transactions from the new block,
+     * some of the transactions in the vector may not be in our mempool.
```
Thank you removed

ismaelsadeeq commented on src/txmempool.cpp:
```diff
@@ -633,33 +625,37 @@ void CTxMemPool::removeConflicts(const CTransaction &tx)
+                .txHeight = it->GetHeight(),
+                .nSizeWithAncestors = it->GetSizeWithAncestors(),
+                .nModFeesWithAncestors = it->GetModFeesWithAncestors()};
+            txs_removed_for_block.push_back(tx_info);
```
Done

ismaelsadeeq commented on src/validationinterface.h:
```diff
@@ -131,9 +132,18 @@ class CValidationInterface {
+     *
+     * Called on a background thread.
+     */
+    virtual void MempoolTransactionsRemovedForConnectedBlock(const std::vector<NewMempoolTransactionInfo>& txs_removed_for_block, unsigned int nBlockHeight) {}
```
Thank you  @glozow
 Now using two structs `RemovedMempoolTransactionInfo` for this callback and `NewMempoolTransactionInfo` for `TransactionAddedToMempool callback.

Since they have similar fields, I created a base struct with all the similar fields so that they can both inherit 79803d3912747eb96153357f1cc33c3f8cf21eb6 and a469d937dbbadedeb58e0d5d1c44ed7881a7ebb9

ismaelsadeeq commented on src/validationinterface.cpp:
```diff
@@ -214,6 +214,16 @@ void CMainSignals::TransactionAddedToMempool(const CTransactionRef& tx, uint64_t
+    auto event = [txs_removed_for_block, nBlockHeight, this] {
+        m_internals->Iterate([&](CValidationInterface& callbacks) { callbacks.MempoolTransactionsRemovedForConnectedBlock(txs_removed_for_block, nBlockHeight); });
+    };
+    ENQUEUE_AND_LOG_EVENT(event, "%s: block height=%s txs=%s", __func__,
```
Fixed

ismaelsadeeq commented on src/kernel/mempool_entry.h:
```diff
@@ -176,4 +176,21 @@ class CTxMemPoolEntry
+     * transactions.
+     */
+    int64_t m_virtual_transaction_size;
+    unsigned int txHeight;
```
Added

ismaelsadeeq commented on src/policy/fees.h:
```diff
@@ -272,6 +274,7 @@ class CBlockPolicyEstimator
     {
         unsigned int blockHeight{0};
         unsigned int bucketIndex{0};
+        CAmount m_fee_per_k;
```
Thank you, this is removed in latest push

ismaelsadeeq commented on src/policy/fees.cpp:
```diff
@@ -634,7 +637,7 @@ bool CBlockPolicyEstimator::processBlockTx(unsigned int nBlockHeight, const CTra
 
 void CBlockPolicyEstimator::processBlock(unsigned int nBlockHeight,
-                                         std::vector<CTransactionRef>& txs_removed_for_block)
+                                         const std::vector<CTransactionRef>& txs_removed_for_block)
```
Fixed

ismaelsadeeq commented:
Thanks for your review @glozow  and @TheCharlatan 
Forced pushed from 6d676c69195dc9032c4558987fb88f4c4b71092c to 09436b21e84cc6cd979fe4942ba1c415c4bc91be
[Compare changes](https://github.com/bitcoin/bitcoin/compare/6d676c69195dc9032c4558987fb88f4c4b71092c..09436b21e84cc6cd979fe4942ba1c415c4bc91be)

- Rework the commits in the order of this [review comment](https://github.com/bitcoin/bitcoin/pull/28368#pullrequestreview-1704427495) which is more cleaner and less verbose.
- Addressed @glozow review comments.
- Cherry picked https://github.com/TheCharlatan/bitcoin/commit/32846141df1a2e110f84f3244d5e98626d379185 commit which removes `policy/fees` from kernel library.

glozow commented on src/txmempool.cpp:
```diff
@@ -496,12 +496,16 @@ void CTxMemPool::removeUnchecked(txiter it, MemPoolRemovalReason reason)
     // even if not directly reported below.
     uint64_t mempool_sequence = GetAndIncrementSequence();
 
+    const uint256 hash = it->GetTx().GetHash();
```
6f40c32fa8dfc742cf5ca7daa28615e066872aec

maybe stop making a copy, and let it be a `Txid`

```suggestion
    const auto& hash = it->GetTx().GetHash();
```

ismaelsadeeq replied:
👍🏾  Done

glozow commented on src/policy/fees.cpp:
```diff
@@ -566,11 +561,26 @@ CBlockPolicyEstimator::CBlockPolicyEstimator(const fs::path& estimation_filepath
-    unsigned int txHeight = entry.GetHeight();
-    uint256 hash = entry.GetTx().GetHash();
+    unsigned int txHeight = tx_info.txHeight;
+    uint256 hash = tx_info.m_tx->GetHash();
```
ac82f58e45421aa5aec1d3e6e992d1c369472a10

```suggestion
    const auto& hash = tx_info.m_tx->GetHash();
```

ismaelsadeeq replied:
Updated, thank you

glozow commented on src/test/fuzz/policy_estimator.cpp:
```diff
@@ -41,7 +41,14 @@ FUZZ_TARGET(policy_estimator, .init = initialize_policy_estimator)
+                const CTxMemPoolEntry& entry = ConsumeTxMemPoolEntry(fuzzed_data_provider, tx);
+                const auto tx_info = NewMempoolTransactionInfo(entry.GetSharedTx(), entry.GetFee(),
+                                                               entry.GetTxSize(), entry.GetHeight(),
+                                                               /* m_from_disconnected_block */ fuzzed_data_provider.ConsumeBool(),
+                                                               /* m_submitted_in_package */ fuzzed_data_provider.ConsumeBool(),
+                                                               /* m_chainstate_is_current */ fuzzed_data_provider.ConsumeBool(),
+                                                               /* m_has_no_mempool_parents */ fuzzed_data_provider.ConsumeBool());
```
you're changing how the fuzz inputs are processed here - maybe have `m_has_no_mempool_parents = ConsumeBool()` and make the others = false/false/true?

ismaelsadeeq replied:
Thanks, its now as you suggested

glozow commented on src/test/policyestimator_tests.cpp:
```diff
@@ -130,14 +158,35 @@ BOOST_AUTO_TEST_CASE(BlockPolicyEstimates)
+                    // Since TransactionAddedToMempool callbacks are generated in ATMP,
+                    // not addUnchecked, we cheat and create one manually here
+                    const int64_t virtual_size = GetVirtualTransactionSize(*MakeTransactionRef(tx));
+                    const NewMempoolTransactionInfo tx_info = NewMempoolTransactionInfo(MakeTransactionRef(tx),
```
ac82f58e45421aa5aec1d3e6e992d1c369472a10
a bit less verbose
```suggestion
                    const NewMempoolTransactionInfo tx_info{MakeTransactionRef(tx),
```

ismaelsadeeq replied:
Done

glozow reviewed (commented):
Maybe add a `SyncWithValidationInterfaceQueue` to the start of the fee estimation RPCs? To avoid spurious failures in `feature_fee_estimation.py`. Also, I can imagine scripts that grab fee estimates once per block as soon as the block arrives - should probably make sure they're fresh.

ismaelsadeeq committed:
rpc: `SyncWithValidationInterfaceQueue` on fee estimation RPC's

This ensures that the most recent fee estimation data is used for the
fee estimation with `estimateSmartfee` and `estimaterawfee` RPC's.

ismaelsadeeq commented on src/Makefile.am:
```diff
@@ -951,7 +951,6 @@ libbitcoinkernel_la_SOURCES = \
   node/chainstate.cpp \
   node/utxo_snapshot.cpp \
   policy/feerate.cpp \
-  policy/fees.cpp \
```
Done

ismaelsadeeq commented:
Force pushed from 09436b21e84cc6cd979fe4942ba1c415c4bc91be to 969f5ec4a8
- Addressed @glozow review comments
- Added a commit that `SyncWithValidationInterfaceQueue` at the start of fee estimation RPC's
- The relationship between `RemovedMempoolTransactionInfo`, `NewMempoolTransactionInfo`, and `TransactionInfo` has been updated to use composition. `TransactionInfo` object is now a member of `RemovedMempoolTransactionInfo` and `NewMempoolTransactionInfo`.

TheCharlatan commented on src/policy/fees.cpp:
```diff
@@ -579,11 +574,26 @@ CBlockPolicyEstimator::CBlockPolicyEstimator(const fs::path& estimation_filepath
+
+void CBlockPolicyEstimator::TransactionRemovedFromMempool(const CTransactionRef& tx, MemPoolRemovalReason reason, uint64_t mempool_sequence)
+{
+    removeTx(tx->GetHash(), false);
```
The bool in `removeTx` is always `false` now, maybe remove it entirely?

ismaelsadeeq replied:
I agree with this point, fixed.
Thank you

TheCharlatan commented on src/policy/fees.cpp:
```diff
@@ -579,11 +574,26 @@ CBlockPolicyEstimator::CBlockPolicyEstimator(const fs::path& estimation_filepath
 CBlockPolicyEstimator::~CBlockPolicyEstimator() = default;
 
-void CBlockPolicyEstimator::processTransaction(const CTxMemPoolEntry& entry, bool validFeeEstimate)
+void CBlockPolicyEstimator::TransactionAddedToMempool(const NewMempoolTransactionInfo& tx, uint64_t mempool_sequence)
```
Nit: Remove the unused argument names in the interface functions?

TheCharlatan replied:
Yes, I was only suggesting to remove the names, e.g. change `uint64_t mempool_sequence` to `uint64_t /*unused*/`.

ismaelsadeeq replied:
thanks I understand,  updated

ismaelsadeeq commented on src/kernel/mempool_entry.h:
```diff
@@ -176,4 +176,33 @@ class CTxMemPoolEntry
+};
+
+struct RemovedMempoolTransactionInfo : public TransactionInfo {
+    RemovedMempoolTransactionInfo(const CTxMemPoolEntry& entry)
```
Fixed

ismaelsadeeq commented on src/txmempool.cpp:
```diff
@@ -496,12 +496,16 @@ void CTxMemPool::removeUnchecked(txiter it, MemPoolRemovalReason reason)
     // even if not directly reported below.
```
I updated the commit message 👍🏾

ismaelsadeeq commented on src/kernel/mempool_entry.h:
```diff
@@ -208,4 +208,33 @@ struct RemovedMempoolTransactionInfo {
+                              const bool chainstate_is_current,
+                              const bool has_no_mempool_parents)
+        : info(tx, fee, vsize, height),
+          m_from_disconnected_block(from_disconnected_block),
```
Updated with your suggestion

ismaelsadeeq commented:
CI failure is unrelated see https://github.com/bitcoin/bitcoin/pull/28905

sedited reviewed (approved):
ACK 7268aaf36bdcc6b466fd16bbf544782571e680bb

hebasto commented:
CI restarted.

ismaelsadeeq commented on src/policy/fees.cpp:
```diff
@@ -579,11 +574,26 @@ CBlockPolicyEstimator::CBlockPolicyEstimator(const fs::path& estimation_filepath
     LOCK(m_cs_fee_estimator);
-    unsigned int txHeight = entry.GetHeight();
-    uint256 hash = entry.GetTx().GetHash();
+    unsigned int txHeight = tx.info.txHeight;
```
Fixed Thanks

ismaelsadeeq commented:
Thanks for your review @TheCharlatan 
Update from 969f5ec to 91504cb
[Compare 969f5ec to 91504cb](https://github.com/bitcoin/bitcoin/compare/969f5ec4a8f823e593bb71cf59c523c49774f069..91504cbe0de2b74ef1aa2709761aaf0597ec66a2)
* Using explicit constructors and list initialization for `NewMempoolTransactionInfo` and `RemovedMempoolTransactionInfo`  struct
* Removed redundant `boolean` argument in `CBlockPolicyEstimator::removeTx`
* Removed unused argument names in the overridden interface `CValidationInterface` functions 
* Update `CBlockPolicyEstimator::processTransaction`  `txHeight` and `feeRate` to `const`'s
* Rebased after #28905 and #28925 to fix CI failure

sedited reviewed (approved):
Re-ACK 91504cbe0de2b74ef1aa2709761aaf0597ec66a2

willcl-ark reviewed (approved):
ACK 91504cbe0de2b74ef1aa2709761aaf0597ec66a2

Glad to see the fee estimator being decoupled in this way and keen to see more development of additional estimators in the future e.g. as per #27995 

I gave the code a review and it seems well thought-out to me at this stage; left a few nits which can be addressed in the case that you re-touch things.

ismaelsadeeq commented:
Thanks for your review @willcl-ark will address nits if there is need to retouch.

achow101 commented:
ACK 91504cbe0de2b74ef1aa2709761aaf0597ec66a2

pablomartin4btc reviewed (commented):
post merge tACK 91504cbe0de2b74ef1aa2709761aaf0597ec66a2

Tested manually using `RPC` commands `estimatesmartfee` and `estimaterawfee` (following cases from `/test/functional/feature_fee_estimation.py`).
